v0/destinations/ga/deleteUsers.js

```diff
@@ -0,0 +1,107 @@
+ * @returns
+ */
+const prepareDeleteRequest = (userAttributes, config, rudderDestInfo) => {
+  const { secret } = rudderDestInfo;
+  // TODO: Should we do more validations ?
+  if (secret && isEmpty(secret)) {
+    throw new ErrorBuilder()
+      .setMessage(
+        // This would happen when server doesn't send "x-rudder-dest-info" header
+        // Todo's in-case this exception happen:
+        // 1. The server version might be an older one
+        // 2. There would have been some problem with how we are sending this header
+        `The "secret" field is not sent in "x-rudder-dest-info" header`
+      )
+      .setStatus(500)
+      .build();
+  }
+  const requests = userAttributes.map(userAttribute => {
+    if (!userAttribute.userId) {
+      throw new ErrorBuilder()
+        .setMessage("User id for deletion not present")
+        .setStatus(400)
+        .build();
+    }
+    // Reference for building userDeletionRequest
+    // Ref: https://developers.google.com/analytics/devguides/config/userdeletion/v3/reference/userDeletion/userDeletionRequest#resource
+    const reqBody = {
+      kind: "analytics#userDeletionRequest",
+      id: {
+        type: "USER_ID",
+        userId: userAttribute.userId
+      }
+    };
+    // TODO: Check with team if this condition needs to be handled
+    if (config.useNativeSDK) {
+      reqBody.propertyId = config.trackingID;
+    } else {
+      reqBody.webPropertyId = config.trackingID;
+    }
+    const headers = {
+      Authorization: `Bearer ${secret?.access_token}`,
+      Accept: "application/json",
+      "Content-Type": "application/json"
+    };
+    return {
+      body: reqBody,
+      headers
+    };
+  });
+  return requests;
+};
+
+/**
+ * This function will help to delete the users one by one from the userAttributes array.
+ *
+ * @param {*} userAttributes Array of objects with userId, emaail and phone
+ * @param {*} config Destination.Config provided in dashboard
+ * @param {Record<string, any> | undefined} rudderDestInfo contains information about the authorisation details to successfully send deletion request
+ * @returns {Array<{ body: any, headers: any }>}
+ */
+const userDeletionHandler = async (userAttributes, config, rudderDestInfo) => {
+  const userDeleteRequests = prepareDeleteRequest(
+    userAttributes,
+    config,
+    rudderDestInfo
+  );
+  await Promise.all(
+    userDeleteRequests.map(async userDeleteRequest => {
+      const response = await httpPOST(
+        GA_USER_DELETION_ENDPOINT,
+        userDeleteRequest.body,
+        {
+          headers: userDeleteRequest.headers
+        }
+      );
+      // process the response to know about refreshing scenario
+      return gaResponseHandler(response);
+    })
+  );
+  return { statusCode: 200, status: "successful" };
+};
+
+const processDeleteUsers = async event => {
+  const { userAttributes, config, rudderDestInfo } = event;
+  executeCommonValidations(userAttributes);
```

Comment:
can we extend this change to other destination implementations too as we have handful of them ?

Reply:
Done. Can you please verify ?